### PR TITLE
use the operation signature as graphql.document in spans

### LINKED ISSUE #2695: Personal data logged out in the json format
**Is your feature request related to a problem? Please describe.**
When using the jons format, the spans are also logged.
In some spans, the graphql variables and document are logged:
```
"span": {
    "apollo_private.field_level_instrumentation_ratio":0.01,
    "apollo_private.graphql.variables":"{\"sku\":\"\"}",
    "graphql.document":"fragment ..."
```

Those values could log some personal data.
As far as I've checked, `"apollo_private.graphql.variables"` does not seem to disclose any value.
But `"graphql.document"` can disclose some personal data:
Ex:
```
{
        users(
          where:"email=\"<removed by me>\"",
        ) {
```

**Describe the solution you'd like**
- either we can remove all values
- or we can configure what can be logged. I guess this is related to https://github.com/apollographql/router/issues/1840

**Describe alternatives you've considered**
No real alternative. The application should not log any personal data.
.

### LINKED ISSUE #2820: Add configuration for what graphql elements are included in span attributes
In particular graphql.document can be very large and also leak information. We need to have a way of disabling this.

When tackling this we need to take a look at how this is configured between metrics and tracing so that we have a common structure.

### PR DESCRIPTION
**Pull request by Geal**
Fix #2695
Fix #2820 

This will prevent personal information (in hardcoded variables) from appearing in logs. A few points to address here:
- getting the the operation signature requires a successful query planning. What do we do if there was an error? We might  need to know which query failed planning
- the operation signature removes variables as expected, but also removes aliases, resulting in fields appearing multiple times, and [reorders fields in some cases](https://github.com/apollographql/router/pull/2703/files#diff-e57da25fd4fd4e20b21b091959ab52c4ab84762e9170ef21cb0e20ed87a54ad0R147)

Maybe here we could benefit from a router side solution instead of using the operation signature. Could we have a query transformation using apollo-compiler to remove indentation and hardcoded variables, but keep aliases and field order? @lrlna WDYT?

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

**Comment by lrlna** [comment]
> Maybe here we could benefit from a router side solution instead of using the operation signature. Could we have a query transformation using apollo-compiler to remove indentation and hardcoded variables, but keep aliases and field order? @lrlna WDYT?

that'd be another good use case for modifying the original source in apollographql/apollo-rs#420

**Comment by BrynCooke** [comment]
`graphql.document` is actually part of the spec:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md.
I understand why we want to limit this, but don't think we should deviate from the spec. Maybe just have an option to disable adding this to spans in the first place.

**Comment by Meemaw** [comment]
Having `graphql.document` as query is extremely useful imo. Could the router just replace inline/hardcoded variables with placeholders? That's what tracing tools usually do, for example for SQL queries etc...

**Comment by BrynCooke** [comment]
We should also consider this option.

**Comment by Geal** [comment]
with [91a3497](https://github.com/apollographql/router/pull/2703/commits/91a3497b2a592051cce714fce096358006d6e19c) there's now an option. Note that the operation signature is still added in all cases: https://github.com/apollographql/router/blob/dev/apollo-router/src/plugins/telemetry/mod.rs#L323-L331

File: docs/source/configuration/tracing.mdx [hunk 1]
```diff
@@ -105,6 +105,16 @@ telemetry:
...
 Using this configuration you will have a response header called `my-trace-id` containing the trace ID. It could help you to debug a specific query if you want to grep your log with this trace id to have more context.
 
+## Recording the GraphQL document
+
+Adds the `graphql.document` and `graphql.operation.name` fields to the supergraph span. Enabled by default.
+
+```yaml title="router.yaml"
+telemetry:
+  tracing:
+    record_graphql_document: true
```
[the remark this answers is not in the document]
Comment (Geal) [reply docs/source/configuration/tracing.mdx:115]: good idea

[the remark this answers is not in the document]
Comment (Geal) [reply docs/source/configuration/tracing.mdx:115]: done in [fbb8b96](https://github.com/apollographql/router/pull/2703/commits/fbb8b96ca9c692ddd4a8516b67e7044f08122973)

**Comment by BrynCooke** [comment]
I think some bits need updating, the changelog is incorrect and we should add some docs.

**Comment by abernix** [comment]
Is the signature we're referring to actually the "stats report key"?

**Comment by Geal** [comment]
Yes  that's the one

**Comment by abernix** [comment]
We should not opaquely use the stats report key as the "graphql document" since it has a very Apollo-specific prefix on it.  In the near term, if we want to use it to solve the stated goal, we should at the very least remove that (that is, the header part before the first `\n`).

**Comment by BrynCooke** [comment]
@Geal Let's close this and tackle in #3226

**Pull request closed without merging**